### PR TITLE
Implement assertfailed

### PR DESCRIPTION
**Pull request by Trickybrain**
Implement assertfailed

### CODE REVIEW

File: python/pytest-selfie/pytest_selfie/plugin.py [hunk 1]
```diff
@@ -32,8 +32,23 @@
...
 
 class FSImplementation(FS):
-    def assert_failed(self, message: str, expected=None, actual=None) -> Exception:
-        raise Exception(message)
+    def assert_failed(self, message, expected=None, actual=None) -> Exception:
+        if expected is None and actual is None:
+            return AssertionError(message)
+
+        expected_str = self._nullable_to_string(expected, "(null)")
+        actual_str = self._nullable_to_string(actual, "(null)")
```
Comment (nedtwigg) [inline python/pytest-selfie/pytest_selfie/plugin.py:40]: The `"(null)"` should be `""`

https://github.com/diffplug/selfie/blob/e6c59053316e3492b3552a5dcf54c9ba598ee7e3/jvm/selfie-runner-kotest/src/commonMain/kotlin/com/diffplug/selfie/kotest/FSOkio.kt#L47-L48

File: python/pytest-selfie/pytest_selfie/plugin.py [hunk 2]
```diff
@@ -32,8 +32,23 @@
...
+
+        if not expected_str and not actual_str and (expected is None or actual is None):
+            return self._comparison_assertion(message, "(null)", "(null)")
```
Comment (nedtwigg) [inline python/pytest-selfie/pytest_selfie/plugin.py:43]: Not quite. We don't want to pass `"(null)", "(null)"`, we want to pass `"(null)"` as the default value if the thing is null

https://github.com/diffplug/selfie/blob/e6c59053316e3492b3552a5dcf54c9ba598ee7e3/jvm/selfie-runner-kotest/src/commonMain/kotlin/com/diffplug/selfie/kotest/FSOkio.kt#L51-L53

File: python/pytest-selfie/pytest_selfie/plugin.py [hunk 3]
```diff
@@ -32,8 +32,23 @@
...
+        return self._comparison_assertion(message, expected_str, actual_str)
+
+    def _nullable_to_string(self, value, on_null):
+        return str(value) if value is not None else on_null
+
+    def _comparison_assertion(self, message, expected, actual) -> Exception:
+        error_message = f"{message} - Expected: {expected}, Actual: {actual}"
+        return AssertionError(error_message)
```
Comment (nedtwigg) [inline python/pytest-selfie/pytest_selfie/plugin.py:51]: In pytest, a comparison assertion should be `assert expected == actual, msg`, see https://github.com/diffplug/selfie/issues/299#issue-2226257195

**Review by nedtwigg — CHANGES_REQUESTED** [review]
Good progress, needs some tweaks.

**Comment by nedtwigg** [comment]
If this feedback is confusing and you would rather work on the `to_match_disk` thing, let me know and I can wrap this up and you could start on the disk integration. Or if you'd like to finish this up that's fine too.

**Review by nedtwigg — APPROVED** [review]
I made a few more tweaks. Do they make sense @Trickybrain?

**Comment by Trickybrain** [comment]
Yes it make more sense this way! Thank you!

**Pull request merged**